# [Calculating GC content](https://rosalind.info/problems/gc/) 

In [1]:
from utils import dna_utils


In [2]:
dna_dict = dna_utils.read_dna_from_fasta_file("../data/rosalind_gc.txt")

In [3]:
def calculate_gc_content(dna_string):
    if not dna_string:
        return "Invalid DNA string"
    num_G = dna_string.count('G')
    num_C = dna_string.count('C')
    return (num_G + num_C )/float(len(dna_string))*100

In [4]:
for id, dna in dna_dict.items():
    print(id, calculate_gc_content(dna))

Rosalind_4618 50.89192025183631
Rosalind_0861 47.855530474040634
Rosalind_7724 50.175438596491226
Rosalind_4692 49.68789013732834
Rosalind_9657 53.0


In [5]:
dna_string = dna_utils.create_dna_string(10000)
print(dna_utils.benchmark(calculate_gc_content, dna_string))
print(dna_utils.benchmark_memory(calculate_gc_content, dna_string))

function calculate_gc_content: time: 0.07295608520507812, dna_length: 10000
('Current Memory Usage: 0.0KB', 'Peak Memory Usage: 0.11328125KB')


In [6]:
dna_string = dna_utils.create_dna_string(100000)
print(dna_utils.benchmark(calculate_gc_content, dna_string))
print(dna_utils.benchmark_memory(calculate_gc_content, dna_string))

function calculate_gc_content: time: 0.6668567657470703, dna_length: 100000
('Current Memory Usage: 125.138671875KB', 'Peak Memory Usage: 912.9072265625KB')


In [7]:
dna_string = dna_utils.create_dna_string(1000000)
print(dna_utils.benchmark(calculate_gc_content, dna_string))
print(dna_utils.benchmark_memory(calculate_gc_content, dna_string))

function calculate_gc_content: time: 6.369113922119141, dna_length: 1000000
('Current Memory Usage: 1014.21875KB', 'Peak Memory Usage: 9368.88671875KB')


In [8]:
dna_string = dna_utils.create_dna_string(10000000)
print(dna_utils.benchmark(calculate_gc_content, dna_string))
print(dna_utils.benchmark_memory(calculate_gc_content, dna_string))

function calculate_gc_content: time: 65.02389907836914, dna_length: 10000000
('Current Memory Usage: 9813.193359375KB', 'Peak Memory Usage: 97803.529296875KB')


In [9]:
dna_string = dna_utils.create_dna_string(100000000)
print(dna_utils.benchmark(calculate_gc_content, dna_string))
print(dna_utils.benchmark_memory(calculate_gc_content, dna_string))

function calculate_gc_content: time: 515.753984451294, dna_length: 100000000
('Current Memory Usage: 97712.787109375KB', 'Peak Memory Usage: 923040.7880859375KB')


# Benchmark Results:
## 1. Generated DNA string (10000 bases)
    function calculate_gc_content: time: 0.06890296936035156, dna_length: 10000
    ('Current Memory Usage: 0.09375KB', 'Peak Memory Usage: 0.18359375KB')

## 2. Generated DNA string (100000 bases)
    function calculate_gc_content: time: 0.6668567657470703, dna_length: 100000
    ('Current Memory Usage: 125.138671875KB', 'Peak Memory Usage: 912.9072265625KB')

## 3. Generated DNA string (1000000 bases)
    function calculate_gc_content: time: 6.369113922119141, dna_length: 1000000
    ('Current Memory Usage: 1014.21875KB', 'Peak Memory Usage: 9368.88671875KB')

## 4. Generated DNA string (10000000 bases)
    function calculate_gc_content: time: 65.02389907836914, dna_length: 10000000
    ('Current Memory Usage: 9813.193359375KB', 'Peak Memory Usage: 97803.529296875KB')

## 5. Generated DNA string (100000000 bases)
    function calculate_gc_content: time: 515.753984451294, dna_length: 100000000
    ('Current Memory Usage: 97712.787109375KB', 'Peak Memory Usage: 923040.7880859375KB')